### ISSUE TITLE
Memory leak caused by message trackTimeout

### ISSUE THREAD

**Issue by miharescic** [dudleycarr/nsqjs#102]
Hej,

I've experienced serious memory leaks under high load (high number of NSQ messages in queue) and traced it down to the `trackTimeout` being set in `message.coffee` file (lines 23-32 ). 

There are 2 things that I suspect could be the reason:
- id message does not timeout, timeout is not cleared when `message.finish()` is called
- compiled javascript code looks suspiciously recursive:

```
(trackTimeout = function() {
      var hard, soft;
      if (_this.hasResponded) {
        return;
      }
      soft = _this.timeUntilTimeout();
      hard = _this.timeUntilTimeout(true);
      _this.timedOut = !soft || !hard;
      if (!_this.timedOut) {
        return setTimeout(trackTimeout, Math.min(soft, hard));
      }
    })();
```

Could you please investigate and verify if this is actually the case?

**Comment by dudleycarr** [comment]
Did you by chance do any memory profiling? That would be a huge help to identify where the issue is. 

As for the track timeout function, I don't think it's the culprit in spite of it being recursive. The number of timeouts for a given message (before hitting the hard timeout) are not that many. Also, the stack size is not increasing since the function is being passed to `setTimeout` which passes the `trackTimeout` function to the event loop.

**Comment by miharescic** [comment]
We did, please see two heap snapshots here:
https://s3-eu-west-1.amazonaws.com/strossle-pfm/20160614T100455.heapsnapshot
https://s3-eu-west-1.amazonaws.com/strossle-pfm/20160614T102435.heapsnapshot

A quick glance over the profile showed a lot of timeouts everywhere, that's where the suspition came from.

**Comment by miharescic** [comment]
I've run some simple tests to reproduce the behavior - 2M nsq messages published, received and finished. 

**Assumption** - until first timeouts clear (`nsqd -msg-timeout="60s"` setting) memory will keep increasing, potentially crash the application.

Images below show different use case memory profiles:

* Original behavior - steep increase in heap size during first 60s
![test-original](https://cloud.githubusercontent.com/assets/18208188/16491206/48b70222-3edd-11e6-9587-124d4b365923.png)
* Behavior without timeout (timeout is not set in code)
![test-notimeout](https://cloud.githubusercontent.com/assets/18208188/16491359/ef8ac804-3edd-11e6-933c-079a1832c9c4.png)
* Behavior with timeout set and cleared when responding (pull-request)
![test-fixed](https://cloud.githubusercontent.com/assets/18208188/16491271/9af17658-3edd-11e6-9abf-3197ba75fb65.png)
* Behavior with original code and `nsqd -msg-timeout="180s" (crashed due to heap out of memory)
![test-original-3m-timeout-crashed-outofmemory](https://cloud.githubusercontent.com/assets/18208188/16491300/b845080a-3edd-11e6-929b-1ba1a1ea8fa6.png)

I've created a pull request that should address this issue: https://github.com/dudleycarr/nsqjs/pull/107

**Comment by dudleycarr** [comment]
@miharescic This is really helpful. Would you mind including the graph of a run for the PR with 180s timeout?

Out of curiosity, what was the max in flight set for this client?

**Comment by miharescic** [comment]
Hej, I'm attaching the nsqd 180s message timeout test with PR code below. max_in_flight was set to 10. Tests were run with node version v6.2.0

![test-pr-nsq-180s-msg-timeout](https://cloud.githubusercontent.com/assets/18208188/16515388/ffb1bca8-3f71-11e6-9f6b-c1e56a660eaa.png)

**Comment by dudleycarr** [comment]
Fixed in #118
